### PR TITLE
Support special characters in wiki page titles.

### PR DESCRIPTION
**Pull request by rczajka**
Redmine wiki pages titles can contain characters which need to be URL-encoded. This means the titles can't be pasted verbatim into the URL templates.

This change introduces a `string.Formatter` subclass which `urrlib.parse.quote`s the argument, and uses that to format the URLs where wiki page titles are used.

### CODE REVIEW

**Comment by maxtepkeev** [comment]
Awesome, looks fantastic, but I would like to have a closer look before merging, will try to do this ASAP.

**Comment by maxtepkeev** [comment]
Finally had some time to take a look, I really like the idea, but having thought about it a bit more I suggest we go a bit further, have a look at the following:

```python
class ResourceQueryFormatter(string.Formatter):
    """
    Quotes query and memorizes all arguments, used during string formatting.
    """
    def __init__(self):
        self.used_kwargs = {}
        self.unused_kwargs = {}

    def check_unused_args(self, used_args, args, kwargs):
        self.used_kwargs = {item: kwargs.pop(item) for item in used_args if item in kwargs}
        self.unused_kwargs = kwargs

    def format_field(self, value, format_spec):
        return quote(super(ResourceQueryFormatter, self).format_field(value, format_spec).encode('utf-8'))


class ResourceQueryStr(str):
    formatter = ResourceQueryFormatter()

    def format(self, *args, **kwargs):
        return self.formatter.format(self, *args, **kwargs)
```

We can wrap all ``query_*`` strings in the ``Resource`` class with ``ResourceQueryStr`` using the ``Registrar`` metaclass and get rid of the direct calls to the custom string formatters in the code.

Also it should be pretty safe to just quote all the ``query_*`` attributes for all the resources and not just ``WikiPage``.

@rczajka Thoughts ?

**Comment by rczajka** [comment]
Looks like a good idea. One more thing I would probably do, just to keep the templating consistent, is then to clearly separate formatting the URLs from plain string manipulation, so change the remaining things like:

```
    @property
    def url(self):
        return '{0}/projects/{1}/issues?query_id={2}'.format(self.manager.redmine.url, …)
```

to things like:

```
    query_xyz = '/projects/{0}/issues?query_id={1}'

    @property
    def url(self):
        return ''.join((self.manager.redmine.url, self.query_xyz.format(…)
```

But as long as it doesn't cough up on weird characters, that's fine by me. :)

**Comment by maxtepkeev** [comment]
@rczajka Makes total sense. I think the url example is the only one that's not in the ``query_*`` family, so we can create a ``query_url`` and move it there.

Do you have time to make the necessary changes ?

If not that's also fine, we can merge this PR as is and I'll adjust the code a bit later. Just please remove the ``.gitignore`` file as it's not supposed to be there.

Thanks.

**Comment by rczajka** [comment]
Sure, I should have same time for that in the next week.

**Comment by maxtepkeev** [comment]
Fantastic. Thanks!

**Comment by rczajka** [comment]
Done.
Haven't touched query strings outside Resource subclasses (specifically, in Watcher and User).

File: redminelib/utilities.py [hunk 1]
```diff
@@ -6,6 +6,14 @@
 import copy
 import string
 import functools
+try:
```
[the remark this answers is not in the document]
Comment (rczajka) [reply redminelib/utilities.py]: Wasn't me! ;)

File: redminelib/resources/base.py [hunk 2]
```diff
@@ -53,6 +55,16 @@ def __new__(mcs, name, bases, attrs):
 
         return registry[name].setdefault('class', cls)
 
+    @staticmethod
+    def update_query_strings(attrs):
+        """
+        Updates all `query_*` string attributes to use ResourceQueryFormatter by defalt.
+        """
+        for k, v in attrs.items():
+            if k.startswith('query_') and isinstance(v, utilities.text_type):
```
[the remark this answers is not in the document]
Comment (rczajka) [reply redminelib/resources/base.py]: My assumption was that you might possibly want to use a different formatter for some query at some point, and then it shouldn't be wrapped in the `ResourceQueryStr` if it's something other than just plain string. But it's not required now and will work the way you suggest too, so I'm changing it your way.

**Comment by maxtepkeev** [comment]
Fantastic. Thanks!

**Pull request merged**